### PR TITLE
Collect and use columns' HasNulls info for RowContainer

### PR DESCRIPTION
**Pull request by zhli1142015**
Collect and use columns' HasNulls info for RowContainer.
1. Collect information on whether each column contains null values.
2. If there is no null values in the column, `extractValuesNoNulls` instead of 
`extractValuesWithNulls` is used to extract results from the row container. This 
would help reduce the unnecessary cost: [nulls buffer memory allocation](https://github.com/facebookincubator/velox/blob/main/velox/exec/RowContainer.h#L929C37-L929C49), 
[null bits checking](https://github.com/facebookincubator/velox/blob/main/velox/exec/RowContainer.h#L942C29-L942C62).

We observed perf gains for several queries from our TPCDS runs
(1TB, 8*8 cores, ABFS) with this change. Listing some of them as below, 
for more details please see: #10776. (As the perf gains are from replacing 
`extractValuesWithNulls` with `extractValuesNoNulls`, some queries' latency 
change is not so noticeable, they are not listed here.)

  | Before | After | PerfGain
-- | -- | -- | --
query14b | 30968 | 29481 | 5%
query47 | 19324 | 18364 | 5%
query65 | 21569 | 20384 | 5%
query72 | 23209 | 22209 | 4%
query95 | 40723 | 39439 | 3%

### CODE REVIEW

File: velox/exec/RowContainer.h [hunk 1]
```diff
@@ -204,7 +204,8 @@ class RowContainer {
             false, // hasNext
             false, // isJoinBuild
             false, // hasProbedFlag
-            false, // hasNormalizedKey
+            false, // hasNormalizedKey,
+            false, // trackColumnsMayHaveNulls,
```
Comment (mbasmanova) [inline velox/exec/RowContainer.h:208]: @zhli1142015 Why do we need this flag? Any particular reason not to track this always?

Comment (zhli1142015) [reply velox/exec/RowContainer.h:208]: This change only affects the hash aggregation, hash join, and sort operators. I introduced this flag to ensure that other operators using RowContainer continue to behave as before. The operators not covered in this change include window, topN, topNRowNumber, and streaming aggregation. Please let me know if I should cover them in the same PR.

Comment (mbasmanova) [reply velox/exec/RowContainer.h]: Having extra flags adds complexity and makes it difficult to reason about the code. Unless there is a reason for some operators to not track nulls of the data stores in RowContainer let's track uniformly.

Comment (zhli1142015) [reply velox/exec/RowContainer.h]: Get it, this flag is removed, thanks.

File: velox/exec/RowContainer.cpp [hunk 2]
```diff
@@ -537,6 +539,45 @@ void RowContainer::store(
...
 }
 
+void RowContainer::storeVector(
+    const DecodedVector& decoded,
+    vector_size_t size,
+    char* const* rows,
+    int32_t column) {
+  auto numKeys = keyTypes_.size();
+  bool isKey = column < numKeys;
+  updateColumnMayHaveNulls(column, decoded.mayHaveNullsRecursive());
```
Comment (mbasmanova) [inline velox/exec/RowContainer.cpp:549]: Why 'recursive'? Is it not enough to check for top-level null?

File: velox/exec/RowContainer.cpp [hunk 3]
```diff
@@ -537,6 +539,45 @@ void RowContainer::store(
...
+  if ((isKey && !nullableKeys_) || !decoded.mayHaveNullsRecursive()) {
+    for (int i = 0; i < size; ++i) {
+      VELOX_DYNAMIC_TYPE_DISPATCH(
```
Comment (mbasmanova) [inline velox/exec/RowContainer.cpp:552]: It seems inefficient to do type dispatch at every loop iteration. Can this be done outside of the loop?

File: velox/exec/SortBuffer.cpp [hunk 4]
```diff
@@ -90,9 +90,8 @@ void SortBuffer::addInput(const VectorPtr& input) {
   for (const auto& columnProjection : columnMap_) {
     DecodedVector decoded(
         *inputRow->childAt(columnProjection.outputChannel), allRows);
-    for (int i = 0; i < input->size(); ++i) {
-      data_->store(decoded, i, rows[i], columnProjection.inputChannel);
-    }
+    data_->storeVector(
```
Comment (mbasmanova) [inline velox/exec/SortBuffer.cpp:93]: This is a nice refactoring. It seems unrelated to nulls optimization though. Would it be possible to extract this into a separate PR to help reduce the size of this PR and allow for easier faster review?

File: velox/exec/RowContainer.h [hunk 5]
```diff
@@ -718,6 +744,15 @@ class RowContainer {
     return keyTypes_;
   }
 
+  inline void updateColumnMayHaveNulls(int32_t columnIndex, bool mayHaveNulls) {
```
Comment (mbasmanova) [inline velox/exec/RowContainer.h:747]: This is a public API? If so, please, document it.

My reading of the code suggests that by default columnsMayHaveNulls_ flags are set to 'false' and the callers are expected to call updateColumnMayHaveNulls to mark columns as maybe having nulls. If callers forget to do that, we'll end up with wrong results. If that's the case, it is very important to clearly document the responsibility of the callers to call this API before storing any values. I wonder why cannot we update these flags when storing values? Would that be more reliable?

Let's also update 'store' APIs to clarify that the caller is expected to call updateColumnMayHaveNulls with proper value before calling 'store'.

File: velox/exec/RowContainer.h [hunk 6]
```diff
@@ -204,7 +204,7 @@ class RowContainer {
             false, // hasNext
             false, // isJoinBuild
             false, // hasProbedFlag
-            false, // hasNormalizedKey
+            false, // hasNormalizedKey,
```
[the remark this answers is not in the document]
Comment (zhli1142015) [reply velox/exec/RowContainer.h]: Removed, thanks.

File: velox/exec/RowContainer.h [hunk 7]
```diff
@@ -302,6 +302,13 @@ class RowContainer {
       char* row,
       int32_t columnIndex);
 
+  /// Stores the 'decoded' into row container at 'column'.
+  void storeVector(
+      const DecodedVector& decoded,
+      vector_size_t size,
```
[the remark this answers is not in the document]
Comment (zhli1142015) [reply velox/exec/RowContainer.h]: I change the `size` parameter to the type `SelectivityVector`, I think it may be more general. Now this fucntion is only called with the all selected vector.

File: velox/exec/RowContainer.h [hunk 8]
```diff
@@ -718,6 +744,15 @@ class RowContainer {
     return keyTypes_;
   }
 
+  inline void updateColumnMayHaveNulls(int32_t columnIndex, bool mayHaveNulls) {
+    columnsMayHaveNulls_[columnIndex] =
```
[the remark this answers is not in the document]
Comment (zhli1142015) [reply velox/exec/RowContainer.h]: I get the below error with above.
`no match for ‘operator|=’ (operand types are ‘std::vector<bool>::reference’ and ‘bool’)`

File: velox/exec/RowContainer.cpp (hunk 3 above)
Comment (zhli1142015) [reply velox/exec/RowContainer.cpp]: Sure, updated thanks.

File: velox/exec/SortBuffer.cpp (hunk 4 above)
Comment (zhli1142015) [reply velox/exec/SortBuffer.cpp:93]: This new function is based on some changes in nulls optimization. And now the change is smaller than before, please let me know if i still need to move it to antoher PR, thanks.

File: velox/exec/RowContainer.h (hunk 5 above)
Comment (zhli1142015) [reply velox/exec/RowContainer.h]: > This is a public API? If so, please, document it.

Added the document, thanks.

> I wonder why cannot we update these flags when storing values? Would that be more reliable?

The reason I avoided doing this is that I thought it would lead to many duplicate updates and degrade performance. However, from my tests, I didn't observe any noticeable impact. Yes, it's better to update this information when modifying null bits. Updated, thanks.

File: velox/exec/RowContainer.h [hunk 9]
```diff
@@ -718,6 +744,17 @@ class RowContainer {
     return keyTypes_;
   }
 
+  /// Updates the specific column's columnHasNulls_ flag, if 'hasNulls' is true.
+  /// columnHasNulls_ flag is false by default.
+  inline void updateColumnHasNulls(int32_t columnIndex, bool hasNulls) {
+    columnHasNulls_[columnIndex] = columnHasNulls_[columnIndex] | hasNulls;
+  }
+
+  /// Returns the specific column's columnHasNulls_ falg.
```
Comment (mbasmanova) [inline velox/exec/RowContainer.h:753]: typo: falg -> flag

File: velox/exec/SortBuffer.cpp (hunk 4 above)
Comment (mbasmanova) [reply velox/exec/SortBuffer.cpp:93]: Yes, let's extract this change into a separate PR. This change is logically separate and much safer than the change for tracking nulls. In case null tracking change causes a regression and needs to be reverted, the revert will be smaller.

File: velox/exec/tests/AggregationTest.cpp [hunk 10]
```diff
@@ -1212,11 +1212,11 @@ TEST_F(AggregationTest, memoryAllocations) {
...
 
-  // Verify memory allocations. Aggregation should make 5 allocations: 1 for the
-  // hash table, 1 for the RowContainer holding accumulators, 3 for results (2
-  // for values and nulls buffers of the grouping key column, 1 for sum column).
+  // Verify memory allocations. Aggregation should make 4 allocations: 1 for the
+  // hash table, 1 for the RowContainer holding accumulators, 2 for results (1
+  // for values of the grouping key column, 1 for sum column).
   planStats = toPlanStats(task->taskStats());
-  ASSERT_EQ(5, planStats.at(aggNodeId).numMemoryAllocations);
+  ASSERT_EQ(4, planStats.at(aggNodeId).numMemoryAllocations);
```
Comment (mbasmanova) [inline velox/exec/tests/AggregationTest.cpp:1219]: Why is this change?

File: velox/exec/RowContainer.h [hunk 11]
```diff
@@ -302,6 +302,13 @@ class RowContainer {
       char* row,
       int32_t columnIndex);
 
+  /// Stores the 'decoded' into row container at 'column'.
+  void storeVector(
+      const DecodedVector& decoded,
+      const SelectivityVector* rows,
```
Comment (mbasmanova) [inline velox/exec/RowContainer.h:308]: Why 'rows' is a pointer and not a reference? Do we allow 'null' rows? If so, let's document this and make sure to test.

That said, if there is no use case for 'rows' where not all rows are selected, let's pass 'size' instead. This is simpler and requires less testing. Furthermore, passing SelectivityVector requires explaining how 'groups' are indexed. Do we store each 'row' in groups[row] or do we store selected rows in groups consecutively.

Finally, 'groups' is not the right name for this parameter. it should be 'rows'. I see how it conflicts with SelectivityVector 'rows'. Once we change it back to 'size' the conflict will go away.

File: velox/exec/tests/AggregationTest.cpp (hunk 10 above)
Comment (zhli1142015) [reply velox/exec/tests/AggregationTest.cpp:1219]: The inputs don't contain nulls, the nulls buffer is not allocated for results after this change. The allocations for result is reduced from 3( group key + related nulls buffer + sum) to 2 (group key + sum).

Comment (mbasmanova) [reply velox/exec/tests/AggregationTest.cpp:1219]: Got it. Makes sense. Thank you for explaining.

File: velox/exec/SortBuffer.cpp (hunk 4 above)
Comment (zhli1142015) [reply velox/exec/SortBuffer.cpp]: Please help to take a look, thanks.
https://github.com/facebookincubator/velox/pull/10812

File: velox/exec/RowContainer.h (hunk 11 above)
Comment (zhli1142015) [reply velox/exec/RowContainer.h]: Updated in the new PR, thanks.

**Comment by zhli1142015** [comment]
[Spark Fuzzer](https://github.com/facebookincubator/velox/actions/runs/10506509651/job/29108721945?pr=10775#logs) test failed due to timeout and there is no output.

**Comment by mbasmanova** [comment]
> [Spark Fuzzer](https://github.com/facebookincubator/velox/actions/runs/10506509651/job/29108721945?pr=10775#logs) test failed due to timeout and there is no output.

Please, investigate. CC: @kgpai @assignUser @rui-mo

**Comment by assignUser** [comment]
@mbasmanova #10739

File: velox/exec/HashTable.cpp [hunk 12]
```diff
@@ -1712,6 +1712,11 @@ void HashTable<ignoreNullKeys>::prepareJoinTable(
     otherTables_.emplace_back(std::unique_ptr<HashTable<ignoreNullKeys>>(
         dynamic_cast<HashTable<ignoreNullKeys>*>(table.release())));
   }
+  for (auto& other : otherTables_) {
+    for (int i = 0; i < rows_->columnTypes().size(); ++i) {
+      rows_->updateColumnHasNulls(i, other->rows()->columnHasNulls(i));
```
Comment (mbasmanova) [inline velox/exec/HashTable.cpp:1717]: This looks like a hack. I assume there are a number of independent RowContainers each with its own set of null flags. Why are we combining these flags?

Comment (mbasmanova) [reply velox/exec/HashTable.cpp:1717]: @zhli1142015 Thank you for clarifying.

> Only the main table is updated in the HashProbe operation and used for extractColumn. 

Here we have a few independent RowContainers and we modify the flag in one of them to reflect the context of all of them. That feels hacky because the flag no longer represents the contents of the RowContainer. I wonder if there is a cleaner way. Can we store this combined flag in the HashTable directly and use in extractColumn?

Comment (zhli1142015) [reply velox/exec/HashTable.cpp]: > Can we store this combined flag in the HashTable directly and use in extractColumn?

Updated, thanks.

> I wonder if there is a cleaner way.

How about we only maintain the null flags in the upper layers of the row container, such as the HashTable and Sort Buffer, and then pass the null flags to the extractColumn function?

Comment (mbasmanova) [reply velox/exec/HashTable.cpp]: > How about we only maintain the null flags in the upper layers of the row container, such as the HashTable and Sort Buffer, and then pass the null flags to the extractColumn function?

That might be a reasonable option. We can also limit this to cases where we need to maintain the combined flag across containers.

Comment (zhli1142015) [reply velox/exec/HashTable.cpp]: I double checked, looks we can't limit above to the case we need to maintian the combined flags across containers. 

I updated the logic according to your suggestion, please help check, thanks.
> Can we store this combined flag in the HashTable directly and use in extractColumn?

**Comment by zhli1142015** [comment]
@mbasmanova , could you help take a look this again? Thanks.

File: velox/exec/RowContainer.h [hunk 13]
```diff
@@ -384,7 +387,25 @@ class RowContainer {
...
+
+  /// Copies the values at 'columnIndex' into 'result' for the 'numRows' rows
+  /// pointed to by 'rows'. If an entry in 'rows' is null, sets corresponding
+  /// row in 'result' to null. If 'columnHasNulls' is false, null free
+  /// optimization would be applied.
+  void extractColumn(
+      const char* const* rows,
+      int32_t numRows,
+      int32_t columnIndex,
+      bool columnHasNulls,
```
Comment (mbasmanova) [inline velox/exec/RowContainer.h:406]: I'm thinking that it makes sense to pass columnHasNulls flag only to static methods, because, when we fetch data from "this" RowContainer then we should let the container to decide whether column has nulls or not.

It would be nice to add comments to static methods that take columnHasNulls flag to explain why they take it and that it is the callers responsibility to ensure the flag is correct.

File: velox/exec/RowContainer.h [hunk 14]
```diff
@@ -718,6 +750,11 @@ class RowContainer {
     return keyTypes_;
   }
 
+  /// Returns the specific column's columnHasNulls_ flag.
```
Comment (mbasmanova) [inline velox/exec/RowContainer.h:753]: Returns true if specified column may have nulls, false otherwise.

File: velox/exec/HashProbe.cpp [hunk 15]
```diff
@@ -79,7 +79,11 @@ void extractColumns(
     }
     child->resize(rows.size());
     table->rows()->extractColumn(
```
Comment (mbasmanova) [inline velox/exec/HashProbe.cpp:81]: Wondering if it would be safer to add extractColumn API to the table itself?

```
table->extractColumn(rows.data(), rows.size(), projection.inputChannel, child)
```

File: velox/exec/HashTable.cpp [hunk 16]
```diff
@@ -1712,6 +1712,13 @@ void HashTable<ignoreNullKeys>::prepareJoinTable(
     otherTables_.emplace_back(std::unique_ptr<HashTable<ignoreNullKeys>>(
         dynamic_cast<HashTable<ignoreNullKeys>*>(table.release())));
   }
+  for (int i = 0; i < rows_->columnTypes().size(); ++i) {
```
Comment (mbasmanova) [inline velox/exec/HashTable.cpp:1715]: Would you add empty lines before and after this block and a small comment to explain what this does?

File: velox/exec/tests/RowContainerTest.cpp [hunk 17]
```diff
@@ -2063,3 +2063,67 @@ TEST_F(RowContainerTest, hugeIntStoreWithNulls) {
...
+
+TEST_F(RowContainerTest, columnHasNulls) {
+  auto rowContainer =
+      makeRowContainer({BIGINT(), BIGINT()}, {BIGINT(), BIGINT()}, false);
+  for (int i = 0; i < rowContainer->columnTypes().size(); ++i) {
+    ASSERT_TRUE(!rowContainer->columnHasNulls(i));
+  }
+
+  const uint64_t kNumRows = 1000;
+  VectorFuzzer fuzzer({.vectorSize = kNumRows, .nullRatio = 0}, pool());
```
Comment (mbasmanova) [inline velox/exec/tests/RowContainerTest.cpp:2075]: Is it necessary to use fuzzer here?

File: velox/exec/tests/RowContainerTest.cpp [hunk 18]
```diff
@@ -2063,3 +2063,67 @@ TEST_F(RowContainerTest, hugeIntStoreWithNulls) {
...
+
+  std::vector<VectorPtr> children;
+  const RowTypePtr& rowType = ROW({BIGINT(), BIGINT(), BIGINT(), BIGINT()});
+  children.reserve(rowType->size());
+  for (auto i = 0; i < rowType->size(); ++i) {
+    if (i % 2 == 0) {
+      children.emplace_back(fuzzer.fuzzFlat(rowType->childAt(i), kNumRows));
+    } else {
+      children.emplace_back(
+          fuzzerWithNulls.fuzzFlat(rowType->childAt(i), kNumRows));
```
Comment (mbasmanova) [inline velox/exec/tests/RowContainerTest.cpp:2088]: There is a slight chance this will not produce any nulls making the test flaky.

File: velox/exec/tests/RowContainerTest.cpp [hunk 19]
```diff
@@ -2063,3 +2063,67 @@ TEST_F(RowContainerTest, hugeIntStoreWithNulls) {
...
+    }
+  }
+
+  auto rowVector = std::make_shared<RowVector>(
```
Comment (mbasmanova) [inline velox/exec/tests/RowContainerTest.cpp:2092]: use makeRowVector from the base class

**Review by mbasmanova — COMMENTED** [review]
@zhli1142015 Overall looks great % a few small questions / comments.

File: velox/exec/tests/RowContainerTest.cpp (hunk 18 above)
Comment (zhli1142015) [reply velox/exec/tests/RowContainerTest.cpp]: I see, updated, thanks.

File: velox/exec/tests/RowContainerTest.cpp (hunk 17 above)
Comment (zhli1142015) [reply velox/exec/tests/RowContainerTest.cpp]: Updated, thanks

File: velox/exec/HashTable.h [hunk 20]
```diff
@@ -400,6 +400,15 @@ class BaseHashTable {
     return offThreadBuildTiming_;
   }
 
+  /// Copies the values at 'columnIndex' into 'result' for the 'numRows' rows
+  /// pointed to by 'rows'. If an entry in 'rows' is null, sets corresponding
+  /// row in 'result' to null.
+  virtual void extractColumn(
+      const char* const* rows,
```
Comment (mbasmanova) [inline velox/exec/HashTable.h:407]: nit: perhaps, combine rows and numRows into folly::Range

File: velox/exec/HashTable.cpp [hunk 21]
```diff
@@ -1712,6 +1712,20 @@ void HashTable<ignoreNullKeys>::prepareJoinTable(
     otherTables_.emplace_back(std::unique_ptr<HashTable<ignoreNullKeys>>(
         dynamic_cast<HashTable<ignoreNullKeys>*>(table.release())));
   }
+
+  // If there are multiple tables, we need to merge the 'columnHasNulls' flags
```
Comment (mbasmanova) [inline velox/exec/HashTable.cpp:1716]: Nice comment. Thanks.

File: velox/exec/HashProbe.cpp [hunk 22]
```diff
@@ -1204,7 +1204,7 @@ void HashProbe::applyFilterOnTableRowsForNullAwareJoin(
     filterTableInput_->resize(numRows);
     filterTableInputRows_.resizeFill(numRows, true);
     for (auto& projection : filterTableProjections_) {
-      tableRows->extractColumn(
+      table_->extractColumn(
```
Comment (mbasmanova) [inline velox/exec/HashProbe.cpp:1207]: Is tableRows variable unused now?

File: velox/exec/RowContainer.h [hunk 23]
```diff
@@ -337,22 +337,30 @@ class RowContainer {
...
       int32_t numRows,
       RowColumn col,
+      bool columnHasNulls,
       vector_size_t resultOffset,
       const VectorPtr& result);
 
   /// Copies the values at 'col' into 'result' for the 'numRows' rows pointed to
   /// by 'rows'. If an entry in 'rows' is null, sets corresponding row in
   /// 'result' to null.
+  /// 'columnHasNulls' indicates whether the 'col' column contains null values.
```
Comment (mbasmanova) [inline velox/exec/RowContainer.h:354]: @param

**Review by mbasmanova — APPROVED** [review]
@zhli1142015 Looks great. Thank you for iterating.

**Comment by zhli1142015** [comment]
https://github.com/facebookincubator/velox/actions/runs/10592606576/job/29352296662?pr=10775

```
FAILED: velox/functions/sparksql/aggregates/tests/CMakeFiles/velox_functions_spark_aggregates_test.dir/MinMaxByAggregationTest.cpp.o 
/opt/homebrew/bin/ccache /Applications/Xcode_15.4.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/c++ -DBOOST_ATOMIC_DYN_LINK -DBOOST_ATOMIC_NO_LIB -DBOOST_CONTEXT_DYN_LINK -DBOOST_CONTEXT_NO_LIB -DBOOST_FILESYSTEM_DYN_LINK -DBOOST_FILESYSTEM_NO_LIB -DBOOST_PROGRAM_OPTIONS_DYN_LINK -DBOOST_PROGRAM_OPTIONS_NO_LIB -DBOOST_REGEX_DYN_LINK -DBOOST_REGEX_NO_LIB -DBOOST_SYSTEM_DYN_LINK -DBOOST_SYSTEM_NO_LIB -DBOOST_THREAD_DYN_LINK -DBOOST_THREAD_NO_LIB -DFMT_SHARED -DFOLLY_HAVE_INT128_T=1 -DGFLAGS_IS_A_DLL=0 -DPROTOBUF_USE_DLLS -DSIMDJSON_THREADS_ENABLED=1 -DVELOX_ENABLE_PARQUET -I/Users/runner/work/velox/velox/. -I/Users/runner/work/velox/velox/velox/external/xxhash -I/Users/runner/work/velox/velox/_build/debug/_deps/xsimd-src/include -I/Users/runner/work/velox/velox/_build/debug/_deps/re2-src -I/Users/runner/work/velox/velox/_build/debug/_deps/duckdb-src/src/include -I/Users/runner/work/velox/velox/_build/debug/_deps/duckdb-src/third_party/fsst -I/Users/runner/work/velox/velox/_build/debug/_deps/duckdb-src/third_party/libpg_query/include -I/Users/runner/work/velox/velox/_build/debug/_deps/duckdb-src/third_party/miniz -I/Users/runner/work/velox/velox/_build/debug/_deps/duckdb-src/third_party/hyperloglog -I/Users/runner/work/velox/velox/_build/debug/_deps/duckdb-src/third_party/fastpforlib -I/Users/runner/work/velox/velox/_build/debug -I/Users/runner/work/velox/velox/_build/debug/_deps/simdjson-src/include -isystem /Users/runner/work/velox/velox/velox -isystem /Users/runner/work/velox/velox/velox/external -isystem /Users/runner/work/velox/velox/_build/debug/_deps/folly-src -isystem /Users/runner/work/velox/velox/_build/debug/_deps/folly-build -isystem /opt/homebrew/include -isystem /opt/homebrew/Cellar/openssl@3/3.3.1/include -isystem /Users/runner/work/velox/velox/velox/external/date/velox/external -isystem /Users/runner/work/velox/velox/_build/debug/_deps/libstemmer/src/libstemmer/include -mcpu=apple-m1+crc -std=c++17 -fvisibility=hidden -D USE_VELOX_COMMON_BASE -D HAS_UNCAUGHT_EXCEPTIONS -Wall -Wextra -Wno-unused        -Wno-unused-parameter        -Wno-sign-compare        -Wno-ignored-qualifiers        -Wno-range-loop-analysis          -Wno-mismatched-tags          -Wno-nullability-completeness -Werror -g -std=gnu++17 -arch arm64 -isysroot /Applications/Xcode_15.4.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX14.5.sdk -fPIE -MD -MT velox/functions/sparksql/aggregates/tests/CMakeFiles/velox_functions_spark_aggregates_test.dir/MinMaxByAggregationTest.cpp.o -MF velox/functions/sparksql/aggregates/tests/CMakeFiles/velox_functions_spark_aggregates_test.dir/MinMaxByAggregationTest.cpp.o.d -o velox/functions/sparksql/aggregates/tests/CMakeFiles/velox_functions_spark_aggregates_test.dir/MinMaxByAggregationTest.cpp.o -c /Users/runner/work/velox/velox/velox/functions/sparksql/aggregates/tests/MinMaxByAggregationTest.cpp
[2989/3330] Building CXX object velox/functions/sparksql/aggregates/tests/CMakeFiles/velox_functions_spark_aggregates_test.dir/LastAggregateTest.cpp.o
[2990/3330] Building CXX object velox/functions/sparksql/aggregates/tests/CMakeFiles/velox_functions_spark_aggregates_test.dir/ModeAggregateTest.cpp.o
ninja: build stopped: subcommand failed.
```

File: velox/exec/RowContainer.h [hunk 24]
```diff
@@ -337,22 +337,32 @@ class RowContainer {
...
   /// corresponding row in 'result' to null.
+  /// @param columnHasNulls indicates whether the 'col' column contains null
+  /// values. If 'columnHasNulls' is false, a null-free optimization will be
+  /// applied. It is the caller's responsibility to ensure this flag is set
+  /// correctly.
   static void extractColumn(
       const char* const* rows,
       int32_t numRows,
       RowColumn col,
+      bool columnHasNulls,
```
Comment (xiaoxmeng) [inline velox/exec/RowContainer.h:348]:
```
nit: s/columnHasNulls/hasNulls/ for short.
```

File: velox/exec/RowContainer.h [hunk 25]
```diff
@@ -337,22 +337,32 @@ class RowContainer {
...
   /// 'result' to null.
+  /// @param columnHasNulls indicates whether the 'col' column contains null
+  /// values. If 'columnHasNulls' is false, a null-free optimization will be
+  /// applied. It is the caller's responsibility to ensure this flag is set
+  /// correctly.
   static void extractColumn(
       const char* const* rows,
       int32_t numRows,
       RowColumn col,
+      bool columnHasNulls,
```
Comment (xiaoxmeng) [inline velox/exec/RowContainer.h:363]: ditto

File: velox/exec/RowContainer.h [hunk 26]
```diff
@@ -1242,12 +1295,20 @@ class RowContainer {
 
   void freeRowsExtraMemory(folly::Range<char**> rows, bool clear);
 
+  // Updates the specific column's columnHasNulls_ flag, if 'hasNulls' is true.
```
Comment (xiaoxmeng) [inline velox/exec/RowContainer.h:1298]: nit:
```
// Updates the specific column's columnHasNulls_ flag if 'hasNulls' is true. 'columnHasNulls_' is false by default.
```

**Review by xiaoxmeng — APPROVED** [review]
@zhli1142015 LGTM % nits. Thanks!

File: velox/exec/RowContainer.h (hunk 24 above)
Comment (zhli1142015) [reply velox/exec/RowContainer.h]: Sure, updated.

File: velox/exec/RowContainer.h (hunk 26 above)
Comment (zhli1142015) [reply velox/exec/RowContainer.h]: Thanks, updated.
